shift_proposal.h

```diff
@@ -2,33 +2,128 @@
 #define SHIFT_PROPOSAL_H
 
 #include <algorithm>
-#include <cassert>
 #include <iterator>
-#include <optional>
+#include <type_traits>
 #include <utility>
 
+template<class I>
+using difference_type_t = typename std::iterator_traits<I>::difference_type;
+
+template<class I>
+using iterator_category_t = typename std::iterator_traits<I>::iterator_category;
+
+template<class I, class Tag, class = void>
+constexpr bool is_category = false;
+template<class I, class Tag>
+constexpr bool is_category<I, Tag, std::enable_if_t<
+    std::is_convertible_v<iterator_category_t<I>, Tag>>> = true;
+
+/// Increment (decrement for negative n) i |n| times or until i == bound,
+/// whichever comes first. Returns n + the difference between i's final position
+/// and its initial position. (Note: "advance" has overloads with this behavior
+/// in the Ranges TS.)
+template<class I>
+constexpr difference_type_t<I> bounded_advance(
+    I& i, difference_type_t<I> n, I const bound)
+{
+    if constexpr (is_category<I, std::bidirectional_iterator_tag>) {
+        for (; n < 0 && i != bound; ++n, void(--i)) {
+            ;
+        }
+    }
+
+    for(; n > 0 && i != bound; --n, void(++i)) {
+        ;
+    }
+
+    return n;
+}
+
 template<class ForwardIt>
-ForwardIt shift_left(ForwardIt first, ForwardIt last, typename std::iterator_traits<ForwardIt>::difference_type n, std::optional<typename std::iterator_traits<ForwardIt>::value_type> filler = std::nullopt)
+ForwardIt shift_left(ForwardIt first, ForwardIt last, difference_type_t<ForwardIt> n)
 {
-    assert(n >= 0);
-    if (n == 0)
+    if (n <= 0) {
         return last;
-    auto shifted_last = std::move(std::next(first, n), last, first);
-    if (filler.has_value())
-        std::fill(shifted_last, last, *filler);
-    return shifted_last;
+    }
+
+    auto mid = first;
+    if (::bounded_advance(mid, n, last)) {
```

Comment:
Perhaps, although it would only matter for non-optimizing compiles (Optimizers will presumably inline `bounded_advance` and propagate the knowledge that `n > 0` from the conditional into its body.).